### ISSUE TITLE
There should be a possibility to send cookies to Chrome.

### ISSUE THREAD

**Issue by ryanzidago** [bitcrowd/chromic_pdf#23]
I've already started this discussion on this [elixir forum post](https://elixirforum.com/t/chromicpdf/29473/12).

What do you exactly mean by "users should be well aware of the security implications"?

In my opinion, having to wait a couple of seconds while the PDF is generated is totally fine (it is what happens when you try to export your document to a PDF format on Google Docs/Google Sheets for example).

I also like to print PDF from an url instead of a file/string because you do not need to write inline CSS (as far as I know the styles are not loaded when you print to PDF from an HTML file or when you render a view as a string in Phoenix).

Setting cookies is easily doable with the [Ferrum library](https://github.com/rubycdp/ferrum#cookies) written in Ruby, as well as with [Puppeteer](https://github.com/puppeteer/puppeteer/blob/v2.1.1/docs/api.md#pagesetcookiecookies). It'll be great if we could also set cookies in Chrome with Elixir, since, to the best of my knowledge, this feature is neither implemented in [pdf-generator](https://github.com/gutschilla/elixir-pdf-generator) nor in [puppeteer-pdf](https://github.com/coletiv/puppeteer-pdf).

**Comment by maltoe** [comment]
@ryanzidago I wrote down what I meant by security implicitations [here in the docs](https://github.com/bitcrowd/chromic_pdf/blob/3b33659e2c918e2c36fe575782fa029256e92a5a/lib/chromic_pdf.ex#L54) - not a big deal, but people should be aware of it.

Regarding cookies:

* Can I ask why you need to set a cookie? i.e. why is there authentication between you and your print target?
* I built a simple solution yesterday which allows setting cookies at boot time (see linked PR). Is this good enough? Unfortunately cookies are stored per session, so to set them programmatically at runtime would mean to either embed it into another command (i.e. as a prefix to the `print_to_pdf` command) or to find a way to send a call to all sessions in the session pool. 🤔

**Comment by ryanzidago** [comment]
@maltoe let's say that you have a website that requires authentication. 

Once the user is authenticated, he/she has the possibility to download as a PDF one of the web page on this same website. 

So when he/she clicks a button to print the actual page, an instance of the chrome browser is launched, but the headless browser can't access the page to print, since it is secured behind some authentication requirement. 

One of the way to bypass this authentication is to reuse the cookie that the user created when he/she authenticated himself/herself on the website, and pass it to chrome when making the request to access the website.

**Comment by maltoe** [comment]
@ryanzidago But the browser and your application should run on the same server / in the same cluster? Why do you not give it its own endpoint that is not authenticated? Feels a bit cumbersome to me to hijack the user session to allow the headless browser to access the same page. Also, if you have user interaction at this point anyway, can't the user just use their normal browser's printing dialog to print the page to PDF?

Regardless, I guess in this case the PR won't help you, at least not in its current state. I'll work on it a bit more when I find the time so that you can pass a cookie to `print_to_pdf/2` (which should be cleared afterwards). Need to come up with a good way to test it, though.

**Comment by ryanzidago** [comment]
What if you have several services? If you have `service A` that generate a PDF and you want it to be accessible for the user on `service A` as well as on `service B`, then it will be more convenient to use Chrome to generate a PDF rather than using the printing dialog. Furhtermore, you do not have to deal with cross-browser compatibility issues since you only use one browser (Chrome) to generate the document, whereas if you use the printing dialog, you will have to deal with a slightly different layout for each browser.

I do not like the idea to have an unauthenticated endpoint. It means that if users find out, they can easily access PDF from other users since authentication is not required. But this solution has been suggested to me many times.

What do you think?

**Comment by maltoe** [comment]
Sorry, "unauthenticated" in the sense of "not-user-authenticated". Not an open endpoint on the internet, but an otherwise secured inter-service endpoint. I mentioned "the same cluster" / VPN because in that case this endpoint (if not exposed to the net) technically wouldn't need authentication. If you say you want two services on the internet where one is a PDF printing service and the other has the URL that  should be printed, you'll of course need some way of authenticating these services. 

However, I'd still argue that in this case you might want to rethink your PDF printing service and instead move your templates over there. Instead of having a generic service that receives a URL and returns a PDF, in my view it should be a service that knows how to print an invoice, a letter, or whatever you're printing, receives parameters for that and returns the PDF. But maybe I'm just not imaginative enough to picture your use-case.

Coming back to the PR: Is boot time setting of cookies enough or would you need it *per PDF printing* then? Do your cookies expire?

**Comment by maltoe** [comment]
@ryanzidago Had some ⏳ and went ahead and implemented it per request, please see #27 

Interface will be

```
cookie = %{
  name: "foo",
  value: "bar",
  domain: "localhost"
}

ChromicPDF.print_to_pdf({:url, "..."}, set_cookie: cookie)
```

Cookies are cleared when PDF is printed. See [Network.setCookie](https://chromedevtools.github.io/devtools-protocol/tot/Network#method-setCookie) for options.

**Comment by ryanzidago** [comment]
@maltoe you are very productive! 
It works like a charm!

Thank you!
